packages/core/src/lifecycle-registry.ts

```diff
@@ -0,0 +1,241 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Binding, ContextView, inject} from '@loopback/context';
+import {CoreBindings, CoreTags} from './keys';
+import {LifeCycleObserver, lifeCycleObserverFilter} from './lifecycle';
+import debugFactory = require('debug');
+const debug = debugFactory('loopback:core:lifecycle');
+
+/**
+ * A group of life cycle observers
+ */
+export type LifeCycleObserverGroup = {
+  /**
+   * Observer group name
+   */
+  group: string;
+  /**
+   * Bindings for observers within the group
+   */
+  bindings: Readonly<Binding<LifeCycleObserver>>[];
+};
+
+export type LifeCycleObserverOptions = {
+  /**
+   * Control the order of observer groups for notifications. For example,
+   * with `['datasource', 'server']`, the observers in `datasource` group are
+   * notified before those in `server` group during `start`. Please note that
+   * observers are notified in the reverse order during `stop`.
+   */
+  groupsByOrder: string[];
+  /**
+   * Notify observers of the same group in parallel, default to `true`
+   */
+  parallel?: boolean;
+};
+
+/**
+ * A context-based registry for life cycle observers
+ */
+export class LifeCycleObserverRegistry implements LifeCycleObserver {
+  constructor(
+    @inject.view(lifeCycleObserverFilter)
+    protected observersView: ContextView<LifeCycleObserver>,
+    @inject(CoreBindings.LIFE_CYCLE_OBSERVER_OPTIONS, {optional: true})
+    protected options: LifeCycleObserverOptions = {
+      parallel: true,
+      groupsByOrder: ['server'],
+    },
+  ) {}
+
+  setGroupsByOrder(groups: string[]) {
+    this.options.groupsByOrder = groups || ['server'];
+  }
+
+  /**
+   * Get observer groups ordered by the group
+   */
+  protected getObserverGroupsByOrder(): LifeCycleObserverGroup[] {
+    const bindings = this.observersView.bindings;
+    const groups = this.sortObserverBindingsByGroup(bindings);
+    if (debug.enabled) {
```

Comment:
It's controlled by `DEBUG` env var. Checking `debug.enabled` to avoid unnecessary calculation if the debug statement uses parameters with expressions.